src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs

```diff
@@ -2,213 +2,1555 @@
+            {
+                if (LambdaUtilities.IsLambdaBodyStatementOrExpression(child))
+                {
+                    continue;
+                }
+
+                if (HasLabel(child))
+                {
+                    yield return child;
+                }
+                else if (_compareStatementSyntax)
+                {
+                    foreach (var descendant in child.DescendantNodes(DescendIntoChildren))
+                    {
+                        if (HasLabel(descendant))
+                        {
+                            yield return descendant;
+                        }
+                    }
+                }
+            }
+        }
+        private bool DescendIntoChildren(SyntaxNode node)
+         => !LambdaUtilities.IsLambdaBodyStatementOrExpression(node) && !HasLabel(node);
 
-        protected const double ExactMatchDist = 0.0;
-        protected const double EpsilonDist = 0.00001;
+        protected internal sealed override IEnumerable<SyntaxNode> GetDescendants(SyntaxNode node)
+        {
+            var rootChildren = (node == _oldRoot) ? _oldRootChildren : (node == _newRoot) ? _newRootChildren : null;
+            return (rootChildren != null) ? EnumerateDescendants(rootChildren) : EnumerateDescendants(node);
+        }
+
+        private IEnumerable<SyntaxNode> EnumerateDescendants(IEnumerable<SyntaxNode> nodes)
+        {
+            foreach (var node in nodes)
+            {
+                if (HasLabel(node))
+                {
+                    yield return node;
+                }
+
+                foreach (var descendant in EnumerateDescendants(node))
+                {
+                    if (HasLabel(descendant))
+                    {
+                        yield return descendant;
+                    }
+                }
+            }
+        }
+
+        private IEnumerable<SyntaxNode> EnumerateDescendants(SyntaxNode node)
+        {
+            foreach (var descendant in node.DescendantNodesAndTokens(
+                descendIntoChildren: ShouldEnumerateChildren,
+                descendIntoTrivia: false))
+            {
+                var descendantNode = descendant.AsNode();
+                if (descendantNode != null && HasLabel(descendantNode))
+                {
+                    if (!LambdaUtilities.IsLambdaBodyStatementOrExpression(descendantNode))
+                    {
+                        yield return descendantNode;
+                    }
+                }
+            }
+
+            bool ShouldEnumerateChildren(SyntaxNode child)
+            {
+                // if we don't want to consider this nodes children, then don't
+                if (!HasChildren(child))
+                {
+                    return false;
+                }
+
+                // Always enumerate the node we were asked about
+                if (child == node)
+                {
+                    return true;
+                }
```

Comment:
Can we use `DescendantNodesAndTokensAndSelf` to the same effect?

Reply:
No, this is slightly different. This isn't including `node` in the enumeration, its ensuring we descend into `node`s children. I'll make the local function static, and pass a lambda in to `DescendantNodesAndTokens` to at least cache the lambda closure.